Validates IndexRequests after ingest pipelines (#91777)

HiDAl opened the pull request:
This code change checks that the document generated/processed by all pipelines is still valid. If the document will be processed by either 1 or 2 pipelines (namely `pipeline` & `finalPipeline`), the validation would be executed only once per document (`IngestService` receives a "list" of requests, basically a bulk request)

closes #91777

Commit:
Add tests to validate the pipeline validation is correctly running

Commit:
fix checkstyle

elasticsearchmachine commented:
Pinging @elastic/es-data-management (Team:Data Management)

joegallo commented on server/src/main/java/org/elasticsearch/ingest/IngestService.java:
```diff
@@ -735,7 +735,9 @@ protected void doRun() {
                     );
 
-                    i++;
+                    validateIndexRequestPostProcessing(requestSlot, indexRequest, onFailure);
```
`executePipelines` is (well, can be) asynchronous, you can't call  `validateIndexRequestPostProcessing` here like this.

joegallo reviewed (changes requested):
Need to address issues with synchronous/asynchronous code, also I'm not absolutely sure the handler invocation is right (but I'm not sure it's wrong either!).

HiDAl commented:
PR waiting merge of #92455

Commit:
Validates IndexRequests after ingest pipelines (#91777)

Validation of data processes by ingest pipelines is not run, so users
could bypass some checks (i.e.: the `_id` of a document is not longer
than 512 bytes, `_source` is not null, etc).

Commit:
convert POJO to Record

Commit:
add an async processor for the multi pipeline tests

HiDAl commented:
I've updated the PR to the latest version with the fix introduced in #92455. Although this fixes the original issue, it has also surfaced a different kinds of errors:

1. In case you send a `_bulk` request dong upserts **and** the index does not exist yet, the validation won't pass:

```
PUT _ingest/pipeline/lowercase-pipeline
{
  "processors": [
    {
      "lowercase": {
        "field": "my_field"
      }
    }
  ]
}

POST _bulk
{"update":{"_id":"1","_index":"test_index","pipeline":"lowercase-pipeline"}}
{"upsert":{"my_field":"UPPER"},"script":{"source":"ctx._source.my_field = ctx._source.my_field.toLowercase()"}}

{
  "took": 0,
  "ingest_took": 2,
  "errors": true,
  "items": [
    {
      "index": {
        "_index": null,
        "_id": null,
        "status": 400,
        "error": {
          "type": "illegal_argument_exception",
          "reason": "pipeline with id [lowercase-pipeline] does not exist"
        }
      }
    }
  ]
}
```

I'll continue trying to tackle this issue.

elasticsearchmachine commented:
Hi @HiDAl, I've created a changelog YAML for you.

HiDAl committed:
Update docs/changelog/92330.yaml
